### PR TITLE
[prebuilds] no prebuilds for inactive repos

### PR DESCRIPTION
**Pull request by svenefftinge**
## Description
This change prevents prebuilds for repositories that have been inactive (no regular workspace starts) for at least a week.
This would have prevented 98835 unnecessary prebuilds in the last 11 days (vs. 28174 prebuilds on repos without projects that are still active).

## Related Issue(s)

See also my comment here https://github.com/gitpod-io/gitpod/issues/9898#issuecomment-1123787585

## Release Notes

```release-note
NONE
```

### CODE REVIEW

File: components/gitpod-db/src/typeorm/workspace-db-impl.ts
Comment (jankeromnes) [inline components/gitpod-db/src/typeorm/workspace-db-impl.ts:372]: Wow, many thanks for this clever fix! 👀

My only worry here is, how do we know that this query on all `d_b_workspace` entries won't be too slow/expensive? (Especially since we'll run it on every received webhook, which seems very frequent.)

Comment (svenefftinge) [reply components/gitpod-db/src/typeorm/workspace-db-impl.ts:372]: This is what EXPLAIN has to say:
Screenshot 2022-05-11 at 17 46 55

Comment (jankeromnes) [reply components/gitpod-db/src/typeorm/workspace-db-impl.ts:372]: Aha, that totally addresses the performance concern then, right? (Or, do we need to somehow test/qualify this?)

**Comment by jankeromnes** [comment]
Briefly testing this, code looks good, so will approve if it works in preview.

File: components/server/ee/src/prebuilds/prebuild-manager.ts
Comment (AlexTugarev) [inline components/server/ee/src/prebuilds/prebuild-manager.ts:369]: Q: Would it make sense to add some caching here? Just in-memory, time based eviction – let's say 1d.

Thinking out loud:
* webhook events can be dense
* checking for a project seems to be a great eviction policy in general

File: components/server/ee/src/prebuilds/prebuild-manager.ts
Comment (jankeromnes) [inline components/server/ee/src/prebuilds/prebuild-manager.ts:201]: Nit: I think this isn't totally accurate (I agree that we want users to create a Project, but the minimal fix is actually just to open a new workspace for that repository).

However, since this error isn't actually shown anywhere in the UI, users can never see it. 😅 So, for something only a Gitpod admin can see while looking into the DB, that's good enough. 💯

File: components/gitpod-db/src/typeorm/workspace-db-impl.ts
Comment (AlexTugarev) [reply components/gitpod-db/src/typeorm/workspace-db-impl.ts:372]: @svenefftinge, the number of rows sent is also a perf issue sometimes. 

If we're just checking for `count === 0` down below, we could start out with a `limit`ed version.

**Review by jankeromnes — APPROVED** [review]
Works like a charm! 🎆 🎉 🎊 

Many thanks @svenefftinge for yet again shipping a clever small fix much faster than the more complicated solution we had in mind to solve this problem (i.e. enforcing projects for all prebuilds). 💯

Adding a hold in case you'd like to request changes @AlexTugarev.

(Will also briefly take a look at the prod DB to double-check there is a key on the workspace `creationTime`.)

/hold

File: components/gitpod-db/src/typeorm/workspace-db-impl.ts
Comment (svenefftinge) [reply components/gitpod-db/src/typeorm/workspace-db-impl.ts:372]: I would expect typeorm to translate to count(*) when using `getCount`. Doesn't it do that?

Comment (AlexTugarev) [reply components/gitpod-db/src/typeorm/workspace-db-impl.ts:372]: 👍🏻 

https://github.com/typeorm/typeorm/blob/62518ae1226f22b2f230afa615532c92f1544f01/src/query-builder/SelectQueryBuilder.ts#L2849

**Comment by geropl** [comment]
@svenefftinge @AlexTugarev  Honstely when looking at the query, I doubt we should send that to prod.

Iterating all regular workspaces from the last 7 days _for every prebuild request_ is quite expensive, and I'm pretty sure will blow our DB budget - and I'd rather not find out. :slightly_smiling_face:

I don't have time to look into this until 10am, so maybe we exclude this one from the rollout.

**Comment by geropl** [comment]
> Would in-memory caching as @AlexTugarev suggested be acceptable?

Ideally I'd like to use an indexed version of `context.repository.cloneUrl` by adding a separate field + index for it (the MySQL way :roll_eyes: ), even if that means we have to clone the content [on insert/update](https://github.com/gitpod-io/gitpod/blob/680932745a14a6453c5bb102ee595e5d9a107619/components/gitpod-db/src/typeorm/workspace-db-impl.ts#L138-L146).

**Comment by svenefftinge** [comment]
Working on it [here](https://github.com/gitpod-io/gitpod/pull/9976).

**Pull request merged**
